rfcs/0000-driver-registry.md

```diff
@@ -0,0 +1,379 @@
+    Userspace {
+        nonce: u32,
+        outgoing: bbq::MpscProducer,
+    },
+}
+```
+
+The driver is responsible for definining the `T` and `U` request/response types associated with a given UUID.
+
+### Using `Uuid`s to identify drivers
+
+There are a lot of different ways to discover drivers used in other operating systems, including using some kind of path, or a type erased file interface.
+
+This proposal takes an approach somewhat similar to Bluetooth's characteristics:
+
+* A UUID is chosen ahead of time to uniquely identify a service
+* Some "official" UUIDs are reserved for MnemOS and built-in drivers
+* A UUID represents a single service interface
+    * Although any driver could implement the UUID, all implementors are expected to have the same interface
+    * The UUID has no concept of versioning. If a breaking change is needed, a new UUID should be used
+
+In practice, any unique/opaque identifier could be used, but a UUID is used for familiarity.
+
+When implementing a driver, the driver would be expected to export its UUID as part of the crate. It's likely the MnemOS project would maintain some registry of "known UUIDs", to attempt to avoid collisions. This could be in a text file, or wiki, etc.
+
+### Using `TypeId`
+
+Although all drivers are accessed by a `KChannel<T>`, which is a heap allocated structure, attempting to hold many channels of different `T`s in the same structure wouldn't work.
+
+Instead, the key:value map would need to type-erase the channel, likely just holding some kind of pointer. Theoretically, the UUID *should* be enough to ensure that the correct `T` matches the expected type, however the cost of getting this wrong (undefined behavior) due to a developer mistake is too high.
+
+In order to avoid this, the `get` function ALSO takes a `TypeId`, which is a unique type identifier provided by the compiler. Since all drivers are statically compiled together, they should all see the same `T` have the same `TypeId`.
+
+This WON'T necessarily work for userspace, or if MnemOS ever gets dynamically loaded drivers, which may be compiled with a different version of Rust, or potentially not Rust at all.
+
+For this reason, non-static users of these drivers will need to use the serialization interface instead. Userspace would send a serialized message to the expected UUID. The kernel would then use the associated deserialization function to deserialize the message, and send it to the associated channel
+
+### Uniqueness of drivers
+
+This scheme requires that only one version of any kind of driver can be registered.
+
+This simplifies things, but could be potentially annoying for certain drivers are expected to have multiple instances, like keyboards, or storage devices.
+
+This proposal sort of ignores that problem, and suggests that those drivers shouldn't be discoverable. Instead, some kind of "higher level" driver that hands out instances of those drivers should be registered.
```

Comment:
so, the expectation here would be something to the effect of...you have a "keyboards" driver that maintains a registry of any device that is a keyboard, and itself depends on individual drivers for specific hardware devices?

Reply:
Yeah, I'd say you go one of two ways:

* A registered "SerialPortManager" driver, that hands out access to ONE of N serial ports on request through it's kproducer
* A registered "KeyboardManager" that takes clients, where all keyboards registered have their output mixed into a single broadcast stream.

Reply:
But really yeah, it isn't really prescribing one way or another, but I guess "consistent driver interface styles" can be done on a later stylistic pass?